File: 07-BACKTRACKING/path-sum.php
<?php

# https://leetcode.com/problems/path-sum/
class Solution
{
    public function hasPathSum(?TreeNode $root, int &$targetSum): bool
    {
        if (!$root) return false;

        $targetSum -= $root->val;

        $isLeafNode = (!$root->left) && (!$root->right);
        if ($isLeafNode && $targetSum == 0) {
            return true;
        }

        if ($this->hasPathSum($root->left, $targetSum) ||
            $this->hasPathSum($root->right, $targetSum)) {
            return true;
        }

        $targetSum += $root->val;
        return false;
    }
}

File: 07-BACKTRACKING/Python/path-sum.py
from typing import *
from TreeNode import TreeNode


# https://leetcode.com/problems/path-sum/
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        if not root:
            return False

        targetSum -= root.val

        isLeafNode = not root.left and not root.right
        if isLeafNode and targetSum == 0:
            return True

        if self.hasPathSum(root.left, targetSum) or \
                self.hasPathSum(root.right, targetSum):
            return True

        targetSum += root.val
        return False


print(Solution().hasPathSum(
    TreeNode.fromArray([1, 2]).print(),
    1
))
